core/src/main/java/jeeves/transaction/TransactionManager.java

```diff
@@ -52,50 +65,100 @@
             definition.setName(name);
             definition.setReadOnly(readOnly);
-            transaction = transactionManager.getTransaction(definition);
-            isNewTransaction = transaction.isNewTransaction();
+            status = transactionManager.getTransaction(definition);
+            isNewTransaction = status.isNewTransaction();
 
             if (isNewTransaction) {
-                Collection<NewTransactionListener> listeners = context.getBeansOfType(NewTransactionListener.class).values();
-                for (NewTransactionListener listener : listeners) {
-                    listener.newTransaction(transaction);
+                try {
+                    fireNewTransaction(context, transactionManager, status);
+                }
+                catch(Throwable t) {
+                    Log.warning(Log.JEEVES, "New transaction:", t);
+                    // warning as we continue with action below
                 }
             }
 
-            result = action.doInTransaction(transaction);
+            result = action.doInTransaction(status);
 
         } catch (Throwable e) {
             Log.error(Log.JEEVES, "Error occurred within a transaction", e);
             if (exception[0] == null) {
                 exception[0] = e;
             }
-            rolledBack = true;
-            doRollback(context, transactionManager, transaction);
+
+            try {
+                doRollback(context, transactionManager, status);
+            }
+            finally {
+                isRolledBack = true;
+            }
         } finally {
             try {
-                if (readOnly) {
-                    doRollback(context, transactionManager, transaction);
-                } else if (!rolledBack && (isNewTransaction || commitBehavior == CommitBehavior.ALWAYS_COMMIT)) {
-                    doCommit(context, transactionManager, transaction);
+                if (readOnly && !isRolledBack) {
+                    try {
+                        doRollback(context, transactionManager, status);
+                    }
+                    finally {
+                        isRolledBack = true;
+                    }
+                } else if (!isRolledBack && (isNewTransaction || commitBehavior == CommitBehavior.ALWAYS_COMMIT)) {
+                    try {
+                        doCommit(context, transactionManager, status);
+                    }
+                    finally {
+                        isCommitted = true;
+                    }
                 }
             } catch (TransactionSystemException e) {
                 if (!(e.getOriginalException() instanceof RollbackException)) {
                     Log.error(Log.JEEVES, "ERROR committing transaction, will try to rollback", e);
-                    doRollback(context, transactionManager, transaction);
+                    if (!isRolledBack) {
```

Comment:
seems this indirection appears three times, maybe four... should maybe take place in a finally clause ? (This, said the need to cascade try/finally reveals maybe a more deep trouble, I don't know)

Reply:
Trying to prevent calling rollback a second time if it was already successfully completed above. It cannot be handled in a finally clause because not all execution paths require rollback (some require commit).

The need to use try/catch repeatedly is defensive coding when calling "extensions" such as listeners that can be added over time. Not every implementation will be written defensively and you want to protect against failure.

Java is just verbose.

Reply:
I begin a refactor in https://github.com/cmangeat/core-geonetwork/tree/transaction_manager
But I cant pr your repository I don't know why.

Reply:
This pull request has the option "Allow edits and access to secrets by maintainers" checked so if you have commit access you should be able to do a direct modification. If it helps you can make a pull request from your branch...